### PR TITLE
Completion of Expose-Commit

### LINKED ISSUE #2: FEATURE - expose commit infomation at /api/systemInfo
# User Story

* As a developer
* I can see information about the latest commit deployed to a deployment on dokku or localhost
* So that I am more aware of which commit is deployed on running code

### Discussion

In this issue, we bring over some code from proj-courses into this code base; code that provides transparency about what version of the code is running in any given deployment.

In the proj-courses code base, we've implemented a feature where you can get the SHA and the commit message for the latest commit that was deployed either on localhost or on dokku.  You access it by going to the endpoint /api/systemInfo.  

Here's what it looks like:

image

Note that this only works if all of the code is deployed correctly in the code base, *and* if the following has been done on the dokku deployment (here, `courses-qa` is the name of the dokku app)

```
  dokku git:set courses-qa keep-git-dir true
```

After setting that, you will need to do a `dokku ps:rebuild appname` for it to take effect.

If that's not set, you'll see something like this:

image

Also, you need to set the env variable: SOURCE_REPO to the correct repo.  This should be done rather than changing the value https://github.com/ucsb-cs156/proj-happycows in the code!

```
dokku config:set organic  SOURCE_REPO=https://ucsb-cs156-s24/proj-organic-s24-5pm-1
```

For the changes you need to make to the code, refer to this PR from proj-courses:

https://github.com/ucsb-cs156/proj-courses/pull/85

### Acceptance Criteria

- [ ] On Dokku, and on localhost, the /api/systemInfo shows the correct commit info about the most recent commit deployed to that deployment

### Implementation Todos

See: https://github.com/ucsb-cs156/proj-courses/pull/85 for hints on how to proceed.

### PR DESCRIPTION
**Pull request by ElijahCanderson**
Displays last commit message and commit Id from github at /api/systemInfo to track development. 

Close #2 
Admin properties have not been implemented for group members but it is not required for checking
Link: https://organic-elijahcanderson-dev.dokku-12.cs.ucsb.edu/api/systemInfo 

Below is the last commit message and the commit Id at the time of upload. 

If wanting to go from home link, you have to manually enter /api/systemInfo

Screen Shot 2024-05-22 at 6 01 48 PM

### CODE REVIEW

File: src/main/java/edu/ucsb/cs156/organic/services/SystemInfoServiceImpl.java [hunk 1]
```diff
@@ -23,13 +23,26 @@ public class SystemInfoServiceImpl extends SystemInfoService {
...
 
   public SystemInfo getSystemInfo() {
     SystemInfo si = SystemInfo.builder()
     .springH2ConsoleEnabled(this.springH2ConsoleEnabled)
     .showSwaggerUILink(this.showSwaggerUILink)
-    .sourceRepo(this.sourceRepo)
+    .sourceRepo("https://github.com/ucsb-cs156-s24/proj-organic-s24-5pm-4")
+    .commitMessage(this.commitMessage)
+    .commitId(this.commitId)
+    .githubUrl(githubUrl("https://github.com/ucsb-cs156-s24/proj-organic-s24-5pm-4", this.commitId))
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/organic/services/SystemInfoServiceImpl.java]:
```suggestion
    .githubUrl(githubUrl( this.sourceRepo, this.commitId))
```
Again, no hard coding.

File: src/test/java/edu/ucsb/cs156/organic/services/SystemInfoServiceImplTests.java [hunk 2]
```diff
@@ -29,7 +30,20 @@ void test_getSystemInfo() {
     SystemInfo si = systemInfoService.getSystemInfo();
     assertTrue(si.getSpringH2ConsoleEnabled());
     assertTrue(si.getShowSwaggerUILink());
-    assertEquals("https://github.com/ucsb-cs156/proj-organic", si.getSourceRepo());
+    assertEquals("https://github.com/ucsb-cs156-s24/proj-organic-s24-5pm-4", si.getSourceRepo());
```
Comment (pconrad) [inline src/test/java/edu/ucsb/cs156/organic/services/SystemInfoServiceImplTests.java]:
```suggestion
       assertEquals("https://github.com/ucsb-cs156/proj-organic", si.getSourceRepo());
```

The default value should be `https://github.com/ucsb-cs156/proj-organic"` in the properties files, and you should be able to override it by setting SOURCE_REPO.

**Review by pconrad — CHANGES_REQUESTED** [review]
I see some things that might help get this working.

You also need to check what's in the properties files... I'll come back and add more comments if I find anything there that needs to be changed.

**Comment by pconrad** [comment]
The property file value is exactly as it should be:
```
app.sourceRepo=${SOURCE_REPO:${$env.SOURCE_REPO:https://github.com/ucsb-cs156/proj-organic}}
```

Leave that as the default value, and only override it to `ttps://github.com/ucsb-cs156-s24/proj-organic-s24-5pm-4/` by setting SOURCE_REPO in either .env or on your dokku instance.

Also, don't forget to do:
```
dokku git:set appname keep-git-dir true
```

on the dokku apps where you run this.  Not needed on localhost since you are running `mvn spring-boot:run` in the directory that has the .git subdirectory already.

**Comment by ElijahCanderson** [comment]
I have already tested without hard coding, and it does not work. The url in the api/system info will not change locally and commit id and message are stuck on unkown.

My implementation in my .env is SOURCE_REPO=https://github.com/ucsb-cs156-s24/proj-organic-s24-5pm-4
But that does not change the URL. I would think this is the issue for the Commit ID and Commit message not working, but configuring it on dokku does change the URL (without hardcoding), yet CommitID and Commit message remain unkown. The only way I have found that the Url changes locally is with hardcoding.
Even if I change it in the properties file, IE, setting my teams repo as the default one, the url remains stuck at the base organic repo locally.

I did make the changes you mentioned as a double check and the issue still persists. Are there other files I should be changing outside of the one's changed in the provided pull request? I have looked and have not found any that stand out.

**Review by pconrad — APPROVED** [review]
LGTM!  Thanks for your patience

**Review by Division7 — DISMISSED** [review]
LGTM 😎 🤙

File: src/main/java/edu/ucsb/cs156/organic/services/SystemInfoServiceImpl.java [hunk 3]
```diff
@@ -23,16 +28,30 @@ public class SystemInfoServiceImpl extends SystemInfoService {
   private boolean showSwaggerUILink;
 
   @Value("${app.sourceRepo}")
-  private String sourceRepo = "https://github.com/ucsb-cs156/proj-organic";
+  private String sourceRepo = "https://github.com/ucsb-cs156-s24/proj-organic-s24-5pm-4";
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/organic/services/SystemInfoServiceImpl.java]: I do need you to change this back.   Ultimately, the code that runs the app is proj-organic.

In terms of figuring out why localhost is different, consider this line:

https://github.com/ucsb-cs156-s24/proj-organic-s24-5pm-4/blob/44a929d1384a0cc88591765d5ae7e340817bb517/src/main/resources/application-development.properties#L13

I'm wondering if the fact that that is there in application-development.properties but not in application-production.properties is part of why it's not working on localhost?

Maybe experiment with that?

**Review by pconrad — CHANGES_REQUESTED** [review]
I know you want to be done with this, and I want you do be (and get the points)... but hear me out...

See comments below...

**Review by pconrad — APPROVED** [review]
LGTM!

**Pull request merged**